go/vt/vttablet/tabletmanager/vreplication/vcopier_copy_all.go

```diff
@@ -0,0 +1,297 @@
+			resp.TableName, len(resp.Fields), len(resp.Rows), resp.Gtid, resp.Lastpk)
+		tableName := resp.TableName
+		gtid = resp.Gtid
+
+		updateRowsCopied := func() error {
+			updateRowsQuery := binlogplayer.GenerateUpdateRowsCopied(vc.vr.id, vc.vr.stats.CopyRowCount.Get())
+			_, err := vc.vr.dbClient.Execute(updateRowsQuery)
+			return err
+		}
+
+		if err := updateRowsCopied(); err != nil {
+			return err
+		}
+		select {
+		case <-rowsCopiedTicker.C:
+			if err := updateRowsCopied(); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+		if tableName != state.currentTableName {
+			if copyWorkQueue != nil {
+				copyWorkQueue.close()
+			}
+			copyWorkQueue = vc.newCopyWorkQueue(parallelism, copyWorkerFactory)
+			if state.currentTableName != "" {
+				log.Infof("copy of table %s is done at lastpk %+v", state.currentTableName, lastpkbv)
+				if err := vc.deleteCopyState(state.currentTableName); err != nil {
+					return err
+				}
+			} else {
+				log.Infof("starting copy phase with table %s", tableName)
+			}
+
+			state.currentTableName = tableName
+		}
+		if !copyWorkQueue.isOpen {
+			if len(resp.Fields) == 0 {
+				return fmt.Errorf("expecting field event first, got: %v", resp)
+			}
+			fieldEvent := &binlogdatapb.FieldEvent{
+				TableName: tableName,
+			}
+			fieldEvent.Fields = append(fieldEvent.Fields, resp.Fields...)
+			tablePlan, err := state.plan.buildExecutionPlan(fieldEvent)
+			if err != nil {
+				return err
+			}
+			pkfields = append(pkfields, pkfields...)
+			buf := sqlparser.NewTrackedBuffer(nil)
+			buf.Myprintf(
+				"insert into _vt.copy_state (lastpk, vrepl_id, table_name) values (%a, %s, %s)", ":lastpk",
+				strconv.Itoa(int(vc.vr.id)),
+				encodeString(tableName))
+			addLatestCopyState := buf.ParsedQuery()
+			copyWorkQueue.open(addLatestCopyState, pkfields, tablePlan)
+		}
+		if len(resp.Rows) == 0 {
+			return nil
+		}
+		// Get the last committed pk into a loggable form.
+		lastpkbuf, merr := prototext.Marshal(&querypb.QueryResult{
+			Fields: pkfields,
+			Rows:   []*querypb.Row{lastpk},
+		})
+		if merr != nil {
+			return fmt.Errorf("failed to marshal pk fields and value into query result: %s", merr.Error())
+		}
+		lastpkbv = map[string]*querypb.BindVariable{
+			"lastpk": {
+				Type:  sqltypes.VarBinary,
+				Value: lastpkbuf,
+			},
+		}
+		log.Infof("copying table %s with lastpk %v", tableName, lastpkbv)
+
+		// Prepare a vcopierCopyTask for the current batch of work.
+		currCh := make(chan *vcopierCopyTaskResult, 1)
+		currT := newVCopierCopyTask(newVCopierCopyTaskArgs(resp.Rows, resp.Lastpk))
+
+		// Send result to the global resultCh and currCh. resultCh is used by
+		// the loop to return results to VStreamRows. currCh will be used to
+		// sequence the start of the nextT.
+		currT.lifecycle.onResult().sendTo(currCh)
+		currT.lifecycle.onResult().sendTo(resultCh)
+
+		// Use prevCh to Sequence the prevT with the currT so that:
```

Comment:
Can you please prepend a paragraph that explains what `prevT` and `nextT` are, in the first place? Also, why do we need to track them all, aren't we running sequentially through all tables?

Reply:
This is again due to https://github.com/vitessio/vitess/pull/10828. I added a note above, linking to that PR. We do need to extract out that code, originally added to the regular vcopier, into a more readable package/set of functions. Didn't want to do that as part of this PR to reduce the regression blast radius of this PR.

Reply:
Again, this is not parallel insertion of tables, but of batches in a single table.